x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -974,4 +1018,66 @@ private boolean verify(SecureString password) {
             return hash != null && cacheHasher.verify(password, hash);
         }
     }
+
+    public static final class ApiKeyDoc {
```

Comment:
Yes, it is intentionally introduced and is more or less a centre piece of this PR, which is about improve the role cache efficiency for API keys, i.e. same role descriptors should only be cached once even when they are from different API keys. The current cache is basically keyed by ApiKey ID, which is unfit for the goal. The obvious choice is to use role descriptors themselves as the key, but it has a few downsides and also not compatible the existing cache. So alternatively, we could use the string representation of role descriptors as the key. To further reduce the memory cost, we actually use sha256 of the string representation as the key. More details can be found [here](https://github.com/elastic/elasticsearch/issues/53939#issuecomment-643021514)

So this new class has roleDescriptors and limitedByRoeDescriptors as `ByteReference` to faciliate the calculation. By storing them as bytes instead of `Map`, it also helps avoid [quite](https://github.com/elastic/elasticsearch/blob/v7.8.0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L338) [a](https://github.com/elastic/elasticsearch/blob/v7.8.0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L411) [few](https://github.com/elastic/elasticsearch/blob/v7.8.0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java#L351-L352) unncessary serialisation and deserialisation. So instead of four de/serialisation, it now only deserialises once. When security index is local to a node, all these de/serialisation operations are a noticable part of the total authentication time (> 20%).

Reply:
Okay, thank you for the explanation!

I would've thought that you can get the `BytesReference` for a source field but haven't checked before asking and it turns out you can't, you must use a parser.